packages/astro/src/i18n/middleware.ts

```diff
@@ -34,20 +36,24 @@ export function createI18nMiddleware(
 			const pathnameContainsDefaultLocale = url.pathname.includes(`/${defaultLocale}`);
 			const isLocaleFree = checkIsLocaleFree(url.pathname, i18n.locales);
 			if (i18n.routingStrategy === 'prefix-other-locales' && pathnameContainsDefaultLocale) {
-				const content = await response.text();
 				const newLocation = url.pathname.replace(`/${defaultLocale}`, '');
 				response.headers.set('Location', newLocation);
 				return new Response(null, {
 					status: 404,
 					headers: response.headers,
 				});
-			}
-			// Astro can't know where the default locale is supposed to be, so it returns a 404 with no content.
-			else if (i18n.routingStrategy === 'prefix-always' && isLocaleFree) {
-				return new Response(null, {
-					status: 404,
-					headers: response.headers,
-				});
+			} else if (i18n.routingStrategy === 'prefix-always') {
+				if (url.pathname === base || url.pathname === base + '/') {
+					return context.redirect(`${joinPaths(base, i18n.defaultLocale)}`);
+				}
+
+				// Astro can't know where the default locale is supposed to be, so it returns a 404 with no content.
```

Comment:
Just reminding ourselves that docs will have to be VERY LOUD about saying that your `defaultLocale` must be set, and even your default languages files need to be under a `/[lang]/` folder.

Since there's no config option to indicate `true` to use this feature and it will just be enabled by default with the object, I'm just checking that the expected behaviour is:

- if no i18n config object is found, then none of this routing strategy applies.
- if an i18 config object is found, THEN all these rules kick in

Reply:
Yes, all the existing tests are passing, and there hasn't been a regression so far; this means that i18n routing rules are only triggered when opting in.

Plus, we have an [extensive configuration validation suite](https://github.com/withastro/astro/blob/9e2801d98fc8c64fd38e109886684b614ecd0452/packages/astro/test/units/config/config-validate.test.js#L81-L162) for `defaultLocale` and `locales`